src/www/ui/api/Controllers/UploadTreeController.php

```diff
@@ -0,0 +1,99 @@
+ */
+
+namespace Fossology\UI\Api\Controllers;
+
+use Fossology\Lib\Dao\ClearingDao;
+use Fossology\Lib\Dao\HighlightDao;
+use Fossology\Lib\Dao\LicenseDao;
+use Fossology\Lib\Data\DecisionScopes;
+use Fossology\Lib\Data\DecisionTypes;
+use Fossology\UI\Api\Models\Info;
+use Fossology\UI\Api\Models\InfoType;
+use Psr\Container\ContainerInterface;
+
+/**
+ * @class UploadTreeController
+ * @brief Controller for UploadTree model
+ */
+class UploadTreeController extends RestController
+{
+
+  /**
+   * @var ContainerInterface $container
+   * Slim container
+   */
+  protected $container;
+
+  /** @var ClearingDao */
+  private $clearingDao;
+
+  /** @var HighlightDao */
+  public $highlightDao;
+
+  /** @var DecisionTypes */
+  private $decisionTypes;
+
+  public function __construct($container)
+  {
+    parent::__construct($container);
+    $this->container = $container;
+    $this->clearingDao = $this->container->get('dao.clearing');
+    $this->licenseDao = $this->container->get('dao.license');
+    $this->highlightDao = $container->get("dao.highlight");
+    $this->decisionTypes = $container->get('decision.types');
+  }
+
+  public function getClearingHistory($request, $response, $args)
+  {
+    $itemId = intval($args['itemId']);
+    $uploadId = intval($args['id']);
+    $uploadDao = $this->restHelper->getUploadDao();
+    $returnVal = null;
+
+    if (!$this->dbHelper->doesIdExist("upload", "upload_pk", $uploadId)) {
+      $returnVal = new Info(404, "Upload does not exist", InfoType::ERROR);
+    } else if (!$this->dbHelper->doesIdExist($uploadDao->getUploadtreeTableName($uploadId), "uploadtree_pk", $itemId)) {
+      $returnVal = new Info(404, "Item does not exist", InfoType::ERROR);
+    }
+
+    if ($returnVal !== null) {
+      return $response->withJson($returnVal->getArray(), $returnVal->getCode());
+    }
+
+    $itemTreeBounds = $uploadDao->getItemTreeBoundsFromUploadId($itemId, $uploadId);
+    $clearingDecWithLicenses = $this->clearingDao->getFileClearings($itemTreeBounds, $this->restHelper->getGroupId(), false, true);
+
+    $data = array();
+    $scope = new DecisionScopes();
+
+    foreach ($clearingDecWithLicenses as $clearingDecision) {
+      $removedLicenses = array();
+      $addedLicenses = array();
+
+      foreach ($clearingDecision->getClearingLicenses() as $lic) {
+        $shortName = $lic->getShortName();
+        $lic->isRemoved() ? $removedLicenses[] = $shortName : $addedLicenses[] = $shortName;
+      }
+      ksort($removedLicenses, SORT_STRING);
+      ksort($addedLicenses, SORT_STRING);
+      $data[] = array(
+        'date' => date('Y-m-d', $clearingDecision->getTimeStamp()),
+        'username' => $clearingDecision->getUserName(),
+        'scope' => $scope->getTypeName($clearingDecision->getScope()),
+        'type' => $this->decisionTypes->getTypeName($clearingDecision->getType()),
+        'addedLicenses' => $addedLicenses,
+        'removedLicenses' => $removedLicenses,
+      );;
```

Comment:
Would suggest to create a new model for this complex data. It will help to make sure the response format is consistent with help of test cases.